### ISSUE TITLE
Duplicate local variable name `bitmap$0` on 2.10/2.11 for some weird lazy vals

### ISSUE THREAD

**Issue by sjrd** [scala-js/scala-js#3180]
Discovered in #3178.

2 partests fail to IR-check: [run/t6272.scala](https://github.com/scala/scala/blob/v2.12.4/test/files/run/t6272.scala) and [run/t3877.scala](https://github.com/scala/scala/blob/v2.12.4/test/files/run/t3877.scala), on 2.11 (and also if ported to 2.10). On 2.12+, the produced IR is correct. Given that they are caused by `lazy` vals, this is not surprising, since scalac significant changed its treatment of lazy vals in 2.12. The errors (for t6272) are:
```
[error] Test.scala(30:7:VarDef): Duplicate local variable name bitmap$0.
[error] Test.scala(2:10:VarDef): Duplicate local variable name bitmap$0.
```
The issue is reproducible in 0.6.x and on master.

The IR looks like this:
```scala
  def init___() {
    this.O::init___();
    mod:LA$<-this;
    val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
    try {
      val x1$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x1$1__p1__sr_IntRef__sr_VolatileByteRef__I(x1$lzy, bitmap$0));
      mod:s_sys_package$.error__T__sr_Nothing$("!")
    } catch (e) {
      val e$2: jl_Throwable = mod:sjsr_package$.wrapJavaScriptException__O__jl_Throwable(e);
      val x2$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x2$1__p1__sr_IntRef__sr_VolatileByteRef__I(x2$lzy, bitmap$0))
    } finally {
      val x3$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x3$1__p1__sr_IntRef__sr_VolatileByteRef__I(x3$lzy, bitmap$0))
    };
    if (mod:sjsr_RuntimeString$.isEmpty__T__Z("")) {
      val x4$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x4$1__p1__sr_IntRef__sr_VolatileByteRef__I(x4$lzy, bitmap$0))
    };
    this.b$1 = true;
    while (this.b__Z()) {
      val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
      val x5$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x5$1__p1__sr_IntRef__sr_VolatileByteRef__I(x5$lzy, bitmap$0));
      this.b$und$eq__Z__V(false)
    }
  }
```
and
```scala
  def method__V() {
    val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
    try {
      val x6$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x6$1__p1__sr_IntRef__sr_VolatileByteRef__I(x6$lzy, bitmap$0));
      mod:s_sys_package$.error__T__sr_Nothing$("!")
    } catch (e) {
      val e$2: jl_Throwable = mod:sjsr_package$.wrapJavaScriptException__O__jl_Throwable(e);
      val x7$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x7$1__p1__sr_IntRef__sr_VolatileByteRef__I(x7$lzy, bitmap$0))
    } finally {
      val x8$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x8$1__p1__sr_IntRef__sr_VolatileByteRef__I(x8$lzy, bitmap$0))
    };
    if (mod:sjsr_RuntimeString$.isEmpty__T__Z("")) {
      val x9$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x9$1__p1__sr_IntRef__sr_VolatileByteRef__I(x9$lzy, bitmap$0))
    };
    var b: boolean = true;
    while (b) {
      val bitmap$0: sr_VolatileByteRef = new sr_VolatileByteRef().init___B(0);
      val x10$lzy: sr_IntRef = new sr_IntRef().init___I(0);
      mod:s_Predef$.println__O__V(this.x10$1__p1__sr_IntRef__sr_VolatileByteRef__I(x10$lzy, bitmap$0));
      b = false
    }
  }
```

**Comment by sjrd** [comment]
Sure enough, `-Xprint:mixin` shows 2 declarations of the same variable:
```scala
    def <init>(): A.type = {
      A.super.<init>();
      {
        @volatile var bitmap$0: runtime.VolatileByteRef = new runtime.VolatileByteRef(0);
        try {
          lazy var x1$lzy: runtime.IntRef = new runtime.IntRef(0);
          scala.this.Predef.println(scala.Int.box(A.this.x1$1(x1$lzy, bitmap$0)));
          scala.sys.`package`.error("!")
        } catch {
          case (_: Throwable) => {
            lazy var x2$lzy: runtime.IntRef = new runtime.IntRef(0);
            scala.this.Predef.println(scala.Int.box(A.this.x2$1(x2$lzy, bitmap$0)))
          }
        } finally {
          lazy var x3$lzy: runtime.IntRef = new runtime.IntRef(0);
          scala.this.Predef.println(scala.Int.box(A.this.x3$1(x3$lzy, bitmap$0)))
        }
      };
      if ("".isEmpty())
        {
          lazy var x4$lzy: runtime.IntRef = new runtime.IntRef(0);
          scala.this.Predef.println(scala.Int.box(A.this.x4$1(x4$lzy, bitmap$0)))
        }
      else
        ();
      A.this.b = true;
      while$1(){
        if (A.this.b())
          {
            @volatile var bitmap$0: runtime.VolatileByteRef = new runtime.VolatileByteRef(0);
            {
              lazy var x5$lzy: runtime.IntRef = new runtime.IntRef(0);
              scala.this.Predef.println(scala.Int.box(A.this.x5$1(x5$lzy, bitmap$0)));
              A.this.b_=(false)
            };
            while$1()
          }
        else
          ()
      };
      ()
    }
  };
```
It could be that it's only the same *name*, but that they have different symbols. But if that were true, our `encodeLocalSym` should have disambiguated them, so that's unlikely. If they indeed have the same symbol, I'll be angry at scalac ...

**Comment by sjrd** [comment]
Yup ... it's the same symbol 😱 

Not sure what to do to work around this.

**Comment by sjrd** [comment]
Minimization:
```scala
object Test {
  def main(args: Array[String]): Unit = {
    method()
  }

  def method(): Unit = {
    lazy val x6 = 6
    println(x6)

    do {
      lazy val x10 = 10
      println(x10)
    } while (false)
  }
}
```
It seems a loop is necessary to trigger the issue. So it's probably related to `LabelDef`s inside scalac.

**Comment by sjrd** [comment]
The code produced by scalac is actually wrong even on the JVM, as it does the wrong thing in some corner cases:
```scala
> scala
Welcome to Scala 2.11.11 (Java HotSpot(TM) 64-Bit Server VM, Java 1.8.0_60).
Type in expressions for evaluation. Or try :help.

scala> def method(): Unit = {
     |   var x1InitCount = 0
     |   lazy val x1 = { x1InitCount += 1; 1 }
     |   assert(x1InitCount == 0)
     |   println(x1)
     |   assert(x1InitCount == 1)
     |
     |   do {
     |     lazy val x2 = 2
     |     assert(x1InitCount == 1)
     |     println(x2)
     |     assert(x1InitCount == 1)
     |   } while (false)
     |
     |   println(x1)
     |   assert(x1InitCount == 1)
     | }
method: ()Unit

scala> method()
1
2
1
java.lang.AssertionError: assertion failed
  at scala.Predef$.assert(Predef.scala:156)
  at .method(<console>:26)
  ... 32 elided
```
The last assertion fails, because the `bitmap$0` has been reset to `0` when entering the `do` loop, because of the redefinition of `bitmap$0`. This causes `x1` to be reinitialized on the last access.

So it does seem that, the way this is compiled on the JVM, it behaves as if `bitmap$0` was declared once at the beginning of the method, and then each `ValDef` declaring `bitmap$0` is like an `Assign` instead.

I guess we could reproduce this behavior in our back-end as follows: in 2.10/2.11, we track which local vars have been declared. If we declare a variable twice in a method, we schedule the IR of that method for a fixup round, where we 1) declare all such variables as mutable at the beginning of the method, initialized with the zero of their type and 2) turn all their `ValDef`s into `Assign`s. It will incur some performance overhead even for sane methods, though, because of the tracking.

@gzm0 WDYT?

**Comment by sjrd** [comment]
Since the JVM compiler is wrong anyway, I am also tempted to call it a won't fix, and blacklist the two faulty partests on 2.11 instead.

**Comment by gzm0** [comment]
I suggest closing as wontfix.

**Comment by sjrd** [comment]
Won't-fixed in dec5c945dd0aabf03bb5d071007d9e21563f8d98.
